pkg/gitreceive/git/sha.go

```diff
@@ -0,0 +1,31 @@
+package git
+
+import (
+	"fmt"
+)
+
+const (
+	// this constant represents the length of a shortened git sha - 8 characters long
+	shortShaIdx = 8
+	fullShaLen  = 40
+)
+
+type ErrGitShaTooShort struct {
+	sha string
+}
+
+func (e ErrGitShaTooShort) Error() string {
+	return fmt.Sprintf("git sha %s was too short", e.sha)
+}
+
+type SHA struct {
+	Full  string
+	Short string
+}
```

Comment:
IMO you're not overthinking, and yes, it's meant to be immutable. It probably makes sense to make these and the fields in `SlugBuilderInfo` non-exported and build exported read-only accessors (as much as I loathe that pattern) as you said, but also so that it's clearer that users should use the `New...` functions. Commit forthcoming